Introduce a regression test suite for chkstat

mgerstner opened the pull request:
Result from my hackweek project. Actually inteded was a rewrite of chkstat, but only the switch to a C++ compilation unit actually took place. The regression test suite is a preparational step for the rewrite to avoid regressions.

mgerstner committed:
chkstat: switch to C++ compilation unit

As an initial step to rewrite chkstat in C++, this change renames the
sources to cpp and changes the compiler toolchain accordingly.

A few changes have been necessary in the code, namely:

- explicit casts from (void*)
- add const to char* where string literals are passed

mgerstner committed:
README.md: add basic readme file that explains the repository's purpose

mgerstner committed:
Makefile: allow to build test version programmatically

Do this using `make CHKSTAT_TEST=1`

mgerstner committed:
tests: introduce a regression test suite for chkstat

This lays the foundation for a test harness that helps catching
regressions both during the upcoming rewrite of chkstat and during later
more incremental changes.

For the start this carries just one simple test case but more complex
ones are about to come.

mgerstner committed:
regtest: support read-only bind mounts, also bind-mount permissions repo

mgerstner committed:
regtest: add helpers for permissions.d per package profiles

mgerstner committed:
regtest: simplify test implementation and readability

mgerstner committed:
regtest: add further test cases that test correct profile application

mgerstner committed:
regtests: add tests for correct application of local profiles

mgerstner committed:
regtest: add another test that checks whether the default profile works

mgerstner committed:
regtest: add a couple of command line parameter test cases

mgerstner committed:
regtest: support caps

mgerstner committed:
regtest: improve error flagging of test cases and introduce warnings

- count errors and warnings implicitly via printError, printWarning
- implicitly determine test result code from number of occured errors

mgerstner committed:
regtest: add capability test case

mgerstner committed:
regtest: sort imports, another piece of rationale

mgerstner committed:
regtest: introduce some basic coloring support to improve readability

mgerstner committed:
regtest: add further helper functions, allow access to main instance

mgerstner committed:
regtest: add test for detection of unexpected parent directory ownership

mgerstner committed:
regtest: add test that checks for rejection of world-writable paths

also reuse some code for parsing chkstat output

mgerstner committed:
regtest: add another test that checks rejection of insecure mixed-owner paths

mgerstner committed:
regtest: add another test case that covers unknown user/group config

mgerstner committed:
regtest: employ subuid and subgid feature in user namespace

This quite complex addition allows test cases that need to chown() or
chgrp() to actually do this from within a user namespace.

The precondition is a correctly configured subuid and subgid file for
the running user account. This is somewhat unfortunate as it complicates
both the code and the requirements for users wanting to run the test
with all features. It's still better then not testing certain logic or
requiring the test to run with full host root privileges.

To be of use for users that can't or don't want to setup subuid and
subgid, the rest of the tests still run in case sub*id support is not
available.

mgerstner committed:
regtest: add test case for user owned symlinks

with sub*id features now available we can also implement this test case
which requires chown().

mgerstner committed:
regtest: add another test that catches when privilege bits are set for special files

mgerstner committed:
regtests: add another test case that catches set*id or caps in world-writable sub-trees

mgerstner committed:
regtest: fix capability assertion helper logic

mgerstner committed:
regtest: make capability setting tests optional

It turns out that only more recent kernels support setting file
capabilities in user namespaces. This needs to be handled in a robust
manner and with an explanation.

This currently affects Leap 15.1.

mgerstner committed:
regtest: add test for symlink targets

mgerstner committed:
chkstat: change and harmonize coding style

using four space indentation, opening bracket on the new line, generally
add a block after control statements except for flow control. Adjust vim
modeline accordingly

maltek commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+		found_file_reject = False
+
+		# we can't evaluate the exit code, it's quite always zero, due
+		# to RPM packaging requirements.
```
But it's not? The RPM scriptlet (in factory) has `|| true` to ignore the exit code, but `chkstat` does do a `exit(1)` when errors were encountered.

If you see something different, it's something is broken, or maybe the `unshare` etc. hide the real exit code?

mgerstner replied:
Hmm, the reasoning I added there was wrong, but for the problem at hand it still proves true:

```
/TestUnexpectedPathOwner/file: has unexpected owner. refusing to correct due to unknown integrity.
```

This situation doesn't cause a non-zero exit code. As various kinds of `safe_open()` errors don't trigger a bad exit code. Therefore I've just adjusted the comment for the moment.

maltek reviewed (changes requested):
I have a few comments, mostly minor/questions.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+				os.symlink(os.readlink(src), dst)
+			elif os.path.isdir(src):
+				os.makedirs(dst)
+				self.bindMount(src, dst)
```
As I commented at another bindMount() spot:

```
# NOTE: would be good mounting this read-only, but trying so   
# fails with                                                       
# `mount: /tmp/permissions: filesystem was mounted, but any subsequent operation failed: Unknown error 5005."
# reason is that a following MS_REC mount fails with EPERM,    
# not quite sure why that is.           
```

mgerstner replied:
As discussed privately, recursive `--rbind -oro` is not possible. I moved and updated the comment in the code to include a better reasoning and I also raise an exception now if a caller wants a recrusive/read-only combination.

mgerstner committed:
regtest: by default automatically (re)build chkstat before testing

This avoids testing unexpected versions of chkstat. It can still be
explicitly overriden by a command line switch --skip-make for special
situations.

mgerstner committed:
Makefile: force remove upon clean target to prevent bogus errors

mgerstner committed:
regtest: bindMount(): explicitly reject read-only recursive mounts

I now better understand why a EPERM was returned here in my previous
tests. Getting this right is to complex for a small regtest suite.
Therefore document it better for future readers.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+				shell = False,
+				stdout = subprocess.PIPE,
+				stderr = subprocess.STDOUT
+			)
```
check_call is a tad simpler, but not really shorter. I replaced it nonetheless, using explicitly subprocess.DEVNULL which was what I wanted here.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+			dst_dir = self.m_fake_root + src
+			try:
+				# symlinks here means "copy them as is"
+				shutil.copytree(src, dst_dir, symlinks = True)
```
I clarified.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+		# use a defined standard umask
+		os.umask(0o022)
+		# make sure we also find admin tools
+		os.environ["PATH"] = os.environ["PATH"] + ":/sbin:/usr/sbin"
```
Yes I also thought about a static PATH, but thought I'm not protecting against an attacker here ;-). It's still more defined with a static PATH so I've adjusted it accordingly.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+			self.getProfilePath(self.m_local_profile)
+		] \
+			+ glob.glob(self.m_permissions_dir + "/*") \
+			+ [self.getProfilePath(profile) for profile in self.m_profiles]:
```
True, the loop gave me a headache while writing it already. I've simplified it.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+
+	def getMode(self, path):
+		s = os.stat(path)
+		return self.extractPerms(s)
```
Looking at the current code I don't see a big benefit in changing the API at the moment. It would rather make things more complex, even though those two tiny functions aren't exactly elegant. I'd rather keep it as is for now.

maltek replied:
Well, I still don't quite see the purpose of this function but I won't block the PR over it.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+
+		if not getcap:
+			self.printWarning(
+				"Couldn't find `getcap` utility, can't fully check capability values"
```
I added a hint to the correct zypper package.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+				parts = line.split(':')
+
+				if len(parts) != 2:
+					# bogus line?
```
Indeed, a strange bit. I adjusted the code to do a `split(':', 1)`

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+		"path1": [ "diagnostic1", "diagnostic2", ... ],
+		...
+		}
+		"""
```
No there are currently no tests for those generic errors.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+			testprofile: (
+				# add a trailing slash to express that we want
+				# a directory here
+				self.buildProfileLine(badfile, 0o640),
```
This was a copy/paste bug. I removed the comment.

mgerstner commented on tests/regtest.py:
```diff
@@ -0,0 +1,2002 @@
+			path, owner, group,
+			# chkstat doesn't like the 0oNNN syntax, so remove the
+			# 'o'
+			oct(mode).replace('o', '')
```
Good idea. I'm using format now.

mgerstner commented:
Thanks Malte for reviewing. Apart from the one minor thing about `extractMode()` and `getMode()` I adjusted everything. I will force-push a rebased version of the PR. I also added a bit of new code to compile `chkstat` automatically before starting testing.
